src/navqueue.c

```diff
@@ -192,6 +192,14 @@ static gboolean goto_file_pos(const gchar *file, gint pos)
 void navqueue_go_back(void)
 {
 	filepos *fprev;
+	GeanyDocument *doc = document_get_current();
+
+	/* If the navqueue is currently at some position A, but the actual cursor is at some other
+	 * place B, we should add B to the navqueue, so that (1) we go back to A, not to the next
+	 * item in the queue; and (2) we can later restore B by going forward.
+	 * (If A = B, add_new_position will ignore it.) */
+	if (doc && doc->file_name)
```

Comment:
@vfaronov I mean dropping the `if (doc)` condition and replacing it with an assertion that checks same condition. This way if it ever got added to the plugin API or something with the button sensitivity changes, a programmer would know they made an error when the assertion fires and tells them during development, but it won't affect the main code path for the release build.